pkg/services/object/get/assembly_v2.go

```diff
@@ -0,0 +1,178 @@
+package getsvc
+
+import (
+	"errors"
+
+	apistatus "github.com/nspcc-dev/neofs-sdk-go/client/status"
+	objectSDK "github.com/nspcc-dev/neofs-sdk-go/object"
+	oid "github.com/nspcc-dev/neofs-sdk-go/object/id"
+	"go.uber.org/zap"
+)
+
+func (exec *execCtx) processV2Split(si *objectSDK.SplitInfo) {
+	if _, set := si.FirstPart(); !set {
+		exec.log.Debug("no first ID found in V2 split")
+		exec.err = errors.New("v2 split without first object ID")
+
+		return
+	}
+
+	linkID, set := si.Link()
+	if set && exec.processV2Link(linkID) {
+		return
+	}
+
+	// fallback to the full chain assembly from the last part
+	prev, set := si.LastPart()
+	if set {
+		exec.processV2Last(prev)
+	}
+}
+
+func (exec *execCtx) processV2Last(lastID oid.ID) {
+	lastObj, ok := exec.getChild(lastID, nil, true)
+	if !ok {
+		exec.log.Debug("failed to read last object")
+		return
+	}
+
+	exec.collectedObject = lastObj.Parent()
+
+	// copied from V1, and it has the same problems as V1;
+	// see it for comments and optimization suggestions
+	if ok := exec.writeCollectedHeader(); ok {
+		if ok := exec.overtakePayloadInReverse(lastID); ok {
+			exec.writeObjectPayload(exec.collectedObject)
+		}
+	}
+}
+
+func (exec *execCtx) processV2Link(linkID oid.ID) bool {
+	linkObj, ok := exec.getChild(linkID, nil, true)
+	if !ok {
+		exec.log.Debug("failed to read link object")
+		return false
+	}
+
+	exec.collectedObject = linkObj.Parent()
+
+	var link objectSDK.Link
+	err := linkObj.ReadLink(&link)
+	if err != nil {
+		exec.log.Debug("failed to parse link object", zap.Error(err))
+		return false
+	}
+
+	if exec.ctxRange() == nil {
+		// GET case
+
+		if exec.writeCollectedHeader() {
+			exec.overtakePayloadDirectly(measuredObjsToIDs(link.Objects()), nil, true)
+			return true
+		}
+
+		exec.log.Debug("failed to write parent header")
+
+		return false
+	}
+
+	// RANGE case
+
+	rng := exec.ctxRange()
+	seekOff := rng.GetOffset()
+	seekLen := rng.GetLength()
+	seekTo := seekOff + seekLen
+	parSize := linkObj.Parent().PayloadSize()
+
+	if seekTo < seekOff || parSize < seekOff || parSize < seekTo {
+		var errOutOfRange apistatus.ObjectOutOfRange
+
+		exec.err = &errOutOfRange
+		exec.status = statusOutOfRange
+
+		return true
+	}
+
+	return exec.rangeFromLink(link)
+}
+
+func (exec *execCtx) rangeFromLink(link objectSDK.Link) bool {
+	children := link.Objects()
+	first, firstOffset, last, lastBound := requiredChildren(exec.ctxRange(), children)
+
+	for i := first; i <= last; i++ {
+		child := children[i]
+
+		var rngPerChild *objectSDK.Range
+		if i == first || i == last {
+			rngPerChild = new(objectSDK.Range)
+
+			if i == first {
+				rngPerChild.SetOffset(uint64(firstOffset))
+			}
+			if i == last {
+				rngPerChild.SetLength(uint64(child.ObjectSize()) - uint64(lastBound+firstOffset))
+			}
+		}
+
+		part, ok := exec.getChild(child.ObjectID(), rngPerChild, false)
+		if !ok {
+			return false
+		}
+
+		if exec.writeObjectPayload(part) {
```

Comment:
shouldn't this be inversion?

Reply:
hm, seems so, strange, fixed

Reply:
oh, that is the reason, yes: https://github.com/nspcc-dev/neofs-node/pull/2716#discussion_r1524678153

returned back